src/Bicep.Core/TypeSystem/KeyVaultSecretReferenceType.cs

```diff
@@ -0,0 +1,14 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Bicep.Core.TypeSystem
+{
+    class KeyVaultSecretReferenceType : ObjectType
```

Comment:
I feel like adding a new type needs **really** good justification as it adds significant complexity in the long run. Is there a reason you can't use the `SecureString` modified string type added in this PR for type checks?

Reply:
I was thinking about this (the additional flag as we spoke) but the amount of places I'd need to add handling this (to forbid from using it as regular string) to not allow using the function was adding even more complexity. And more important - the error thrown was saying about trying to assign string to string, which was odd.

Reply:
I'm not convinced at all that enforcing that `getSecret` method only is allowed with secure parameters actually belongs in the core compiler logic. Feels like this should be the job of the linter to catch. (Of course, I would want that linter rule to default to error and be run by default.)

Reply:
I think it should, as we change what is being compiled. Instead a `value` object we emit `reference`. It’s not just a runtime value change, it’s changing compiler logic.

Reply:
The transpiling of `value` into `reference` must be in the core compiler for sure. I'm talking about blocking usage of `getSecret()` with non-secure parameters. That part doesn't have to be in the core compiler and could be delegated to a linter rule.